WIP: Delete Builds associated with a BC after it is deleted

mnagy opened the pull request:
I'm still missing tests, but would like feedback as soon as possible. I've used a different approach than we do with Deployments and DeploymentConfigs, since that one is susceptible to the config being deleted with Deployments remaining.

I've also managed to do without introducing a phase field into the BC. Instead, it is sufficient to enable graceful deletion in the BC registry strategy. This will mark the BC with a deletion timestamp which I'm using to disable any further Builds to be instantiated. If any Build fails to get deleted, the BC controller will return error, so the whole thing gets retried later.

@bparees wanna review?

bparees commented on pkg/build/controller/config_controller.go:
```diff
@@ -6,19 +6,38 @@ import (
+				glog.Errorf("Cannot delete Build %s/%s: %v", build.Namespace, build.Name, err)
+			}
+		}
+		return err
```
is this error type going to be handled as a Retryable error or a fatal error by the RetryController that wraps the BuildConfigController?  I'm guessing we do want to retry this.

Also I don't love that you're just returning the last error, it might be better to create a custom error type that embeds an array of errors for all the failed deletions, particularly since they might have different reasons.

mnagy replied:
Yes, all errors from BC controller are handled as retryable. BTW, just discovered that 60 retries are maximum.

Will fix the second issue.

bparees replied:
yeah it'll retry 60 times, but ultimately the list/watch loop will re-list all the buildconfigs so this will get processed again as long as the DeletionTimestamp is still set, and retried another 60 times.

bparees commented on pkg/build/generator/generator.go:
```diff
@@ -170,6 +170,10 @@ func (g *BuildGenerator) Instantiate(ctx kapi.Context, request *buildapi.BuildRe
 	}
 
+	if bc.DeletionTimestamp != nil {
+		return nil, fmt.Errorf("Cannot generate Build from BuildConfig %s/%s because it is being deleted", bc.Namespace, bc.Name)
```
this needs to be a status error since it goes back out over the rest api, maybe InternalError: https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/errors/errors.go#L236

bparees commented on pkg/build/registry/buildconfig/strategy.go:
```diff
@@ -80,7 +80,9 @@ func SelectableFields(buildConfig *api.BuildConfig) fields.Set {
-	return false
+	options.GracePeriodSeconds = new(int64)
+	*options.GracePeriodSeconds = 0
+	return true
```
what is this actually accomplishing in terms of the buildconfig delete flow/behavior?

bparees commented on pkg/build/controller/config_controller.go:
```diff
@@ -6,19 +6,38 @@ import (
 func (c *BuildConfigController) HandleBuildConfig(bc *buildapi.BuildConfig) error {
 	glog.V(4).Infof("Handling BuildConfig %s/%s", bc.Namespace, bc.Name)
 
+	if !bc.DeletionTimestamp.IsZero() {
```
what is setting the DeletionTimestamp?  and when does the BuildConfig actually get deleted?

mnagy replied:
Relevant function in upstream is here https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/generic/etcd/etcd.go#L351 and another function called by it here https://github.com/kubernetes/kubernetes/blob/master/pkg/api/rest/delete.go#L42.

While researching when exactly the BC gets deleted, I discovered immediately. I'll look into it tomorrow. I think I might need to add that phase after all. :disappointed:

bparees replied:
don't you need to return a non-zero value from CheckGracefulDelete?

but i think gracefuldelete will be a problematic way to solve this since we do not want to delete the buildconfig as long as there are any builds that weren't deleted yet.  so you don't want the grace period to expire and k8s to delete the buildconfig while you're still trying to delete builds, or you'll never delete the builds.

I don't know that you need an additional phase, but i think you do need to mark the buildconfig as deletepending (possibly with the timestamp) and have this controller actually delete the buildconfing once all the builds are successfully deleted.

you'll need to make sure HandleBuildConfig doesn't get called on the delete event, or course, or otherwise do something to ensure you don't create a loop.

mnagy replied:
We could set the grace period as very large. Then, in the BC controller, once we know all builds are deleted, we could shorten the period. I'm not 100% sure this could be done from the controller, but I'll go check.

mnagy commented:
Reworked. I added a CanDelete boolean flag to the BuildConfig status. Instead of just changing the strategy, I've implemented the delete function itself. The delete function will change the timestamp, but won't delete until the CanDelete flag is set. It is set by the BC controller, after all builds were deleted.

bparees commented on pkg/build/controller/config_controller.go:
```diff
@@ -5,20 +5,59 @@ import (
 	glog.V(4).Infof("Handling BuildConfig %s/%s", bc.Namespace, bc.Name)
 
+	if !bc.DeletionTimestamp.IsZero() {
+		if bc.Status.CanDelete {
```
when would this ever happen?

mnagy replied:
Yes, I've seen this. It gets called after the BC update few lines down.

bparees replied:
ok that makes sense.  but i think you can get rid of this field entirely.  all you need is the timestamp to indicate that the deletion has been requested.  every time handle build gets run, if the timestamp is set, you attempt to delete all the child builds.  and once all the child builds are deleted, you attempt to delete the buildconfig.

setting CanDelete is really just saving you the effort of checking if there are any child builds in the case where you already deleted all the child builds, but then failed to delete the buildconfig.

mnagy replied:
But in the end, the etcd REST Delete handler gets called. And it has no way of knowing that builds were deleted and that it can proceed.

bparees replied:
yeah sorry i didn't have the full flow in my head.  you're correct.

bparees commented on pkg/build/registry/buildconfig/etcd/etcd.go:
```diff
@@ -47,3 +51,32 @@ func NewStorage(s storage.Interface) *REST {
+		now := unversioned.Now()
+		bc.DeletionTimestamp = &now
+		bcObj, _, err = r.Update(ctx, bc)
+		return bcObj, err
```
you're not really using this field as intended anymore, right? (this logic is no longer based on the upstream GracefulDelete path) so i think it would be better to introduce a new DeletePending boolean field, rather than abusing the DeletionTimestamp field which could result in unexpected behavior.

mnagy replied:
Well, it kind of is. I based this on the namespace deletion logic: https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/namespace/etcd/etcd.go#L88 with the exception that we're not using a Phase to mark it as terminating, since I thought it would be redundant. I'm not sure why namespaces need it. Given that namespaces use it, I don't think it would be useful to add another field.

bparees replied:
ok, i guess there's precedent, though i'm still a little wary.

bparees commented:
i think this looks ok other than needing tests.

also be aware that clayton just added a new instantiatebinary endpoint that can create builds from buildconfigs, so once that merges you'll need to update that as well to check if the buildconfig is being deleted.

bparees commented on pkg/build/registry/buildconfig/etcd/etcd.go:
```diff
@@ -47,3 +51,32 @@ func NewStorage(s storage.Interface) *REST {
+
+	if !bc.Status.CanDelete {
+		err = kapierrors.NewConflict("BuildConfig", fmt.Sprintf("%s/%s", bc.Namespace, bc.Name), fmt.Errorf("The system is ensuring all Builds instantiated from this BuildConfig are removed. Upon completion, this BuildConfig will automatically be purged by the system."))
+		return nil, err
```
is this really an error?  the scenario here is that delete was called against the buildconfig once, and now it was called again.    that's not really a problem.  why not just no-op it?

mnagy replied:
What do you mean by no-oping it? I think it is kind of an error, because the BC did not get deleted. And I'm not sure, but I suspect it'll get retried thanks to us returning error. Also, the namespace code does the same.

bparees replied:
well it may not have gotten deleted because yet because it's still in the process of being deleted.  you almost certainly don't want to retry it.

the point is, now that the buildconfig has a DeletionTimestamp, the buildconfig controller is going to attempt to delete it every time a resync interval hits.

let's go through the flow here:
1) buildconfig gets marked with a DeletionTimestamp
2) buildcontroller sees the buildconfig update
3) buildcontroller starts deleting child builds
4) if during this time, delete gets called again, it's not an error, nothing needs to happen at all.  the DeletionTimestamp is already set, but it's not an error to call delete on it again.
5) buildcontroller finishes deleting children, sets CanDelete on the BC and calls delete on the buildconfig
6) buildcontroller sees the update to CanDelete, ignores it.
7) REST api sees delete request, notes CanDelete is true, deletes it.

if something goes wrong in step 5, such that the BC never gets updated to CanDelete=true, or if the REST call to delete it fails, it'll get retried during the next list/sync loop in which the controller will process all the buildconfigs again and see a buildconfig that has a DeletionTimestamp, at which point it should pick up at step 2/3.

the problem scenario is when the buildcontroller tries to delete the buildconfig, but gets an error.  As you noted, that's going to result in a retry loop where HandleBuildConfig gets called again with the buildconfig object.  That's fine to a point, but that only retries for a minute or so and then gives up.  At that point you're going to be left with a buildconfig that has a DeletionTimestamp and has CanDelete=true, and it's never going to get deleted unless someone manually invokes the rest api to delete it again.

bparees replied:
the operative mentality here is "eventually consistent", you should be structuring the logic such that states get updated in such a way that you can reprocess the object over and over again, slowly progressing it towards the desired state, even if errors occur at any point in the process.  You should be able to start processing again from any state and make forward progress.  and you should never abandon something in an incomplete state, such as "CanDelete=true but the object still exists 6 hours later"

mnagy replied:
This was a bit long time ago, but.. I think I remember testing the scenario and the result was that the Deletion method was called repeatedly, precisely because the deletion timestamp was set. I will test this scenario again after I rebase.

bparees replied:
it's still not an error right?  if i call delete twice on the same bulidconfig, the first time i'm going to get a successful response and the second time i'm going to get an error, which doesn't make sense since there's really no error from an end-user perspective.  the buildconfig is still successfully being deleted just like it was after the first call.

kargakis commented on pkg/build/util/util.go:
```diff
@@ -74,3 +75,9 @@ func NameFromImageStream(namespace string, ref *kapi.ObjectReference, tag string
+// ConfigSelector returns a label Selector which can be used to find all
+// builds for a BuildConfig.
+func ConfigSelector(name string) labels.Selector {
+	return labels.Set{buildapi.BuildConfigLabel: name}.AsSelector()
```
@deads2k why are we using a different label here: https://github.com/openshift/origin/blob/663ca6f814b1130a9eadd5de7a0a2fa420e4eeaa/pkg/cmd/util/clientcmd/factory.go#L204

bparees replied:
we just deprecated the "buildconfig" label in favor of an "openshit.io/build-config.name" label.  I think this code hasn't been rebased since the label was changed.  Really anywhere we're looking at the buildconfig label now, we need to look for both the old label and the new label, until we can fully drop support for the old label.

@php-coder should have been a part of your pull introducing the new label, sorry i didn't think of it at the time.  can you do a follow up?

mnagy replied:
Not knowing 100% what's going on -- can I just assume I can continue using the `buildapi.BuildConfigLabel`?

kargakis replied:
@bparees this means that older builds need to add the new label or a selector with both labels won't pick them up, right?

bparees replied:
@kargakis existing code needs to check for both labels.

at a minimum existing code needs to check for the deprecated label since that will continue to be available, but in order to be able to eventually remove the old label, we need all the existing code to look for either label.

bparees replied:
> can I just assume I can continue using the buildapi.BuildConfigLabel?

no.  as @kargakis said, old builds won't have the BuildConfigLabel, they'll have the Deprecated one.  Hence my comment that you need to check for both.

kargakis replied:
> @kargakis existing code needs to check for both labels.

I cannot see how this can be done w/o an OR selector... Older builds won't be picked up by newer versions of controllers/clients if you require both labels.

bparees replied:
@kargakis that's what i'm suggesting.  was that not clear?

php-coder replied:
>should have been a part of your pull introducing the new label, sorry i didn't think of it at the time. can you do a follow up?

@bparees Ok, I'll update code. Am I right that I just need to add new commit to the existing branch and reopen original ticket? Or it will be a completely new PR?

php-coder replied:
>Am I right that I just need to add new commit to the existing branch and reopen original ticket? Or it will be a completely new PR?

Just for the record, this is an answer from @bparees  in IRC: `yes, new PR`

mnagy commented:
Rebased, added unit tests and extended tests.

tagging @openshift/api-review

liggitt commented:
If I understand correctly, this is adding cascade deletion in the API, in a way that cannot be opted out of. While I like having cascade delete in the API, there are cases where deletion followed by recreation is done (`oc replace --force`) and I don't think we would want to cascade.

liggitt commented:
I think we have a similar issue for deployment configs... if I force recreate a deployment config, I would NOT expect all my current deployments to be killed

bparees commented:
oc replace actually has a --cascade=false option, so there's definitely an assumption here that we'd respect such an option.

i'm curious how that is done though... it seems to conflict w/ eventually consistent goals.  We need the ability to delete any build that has a reference to a buildconfig that no longer exists.  Not sure how we solve the problem of "the buildconfig doesn't exist right now but it will again in a second, so don't do anything"

liggitt commented:
nothing server-side is cascading today upstream... it's all client side (which I dislike, but allows `--cascade=false` to work)

bparees commented:
hm.  well short of adding a "cascade=false" field to the buildconfig definition, and then updating the client side logic to set that flag on the BC prior to calling delete on the BC, i don't think we can have server side behavior like this and support client side "cascade=false".

so is this a showstopper in your mind?

bparees commented on test/extended/builds/remove_buildconfig.go:
```diff
@@ -0,0 +1,68 @@
+			})
+			if err == wait.ErrWaitTimeout {
+				g.Fail("timed out waiting for builds to clear")
+			}
```
you should also check that the BC itself went away at the end, and that it did not go away before the builds went away.

liggitt commented:
The current behavior of the DC controller makes `oc replace --force` super dangerous. We shouldn't make another resource behave that way unconditionally if we can help it.

bparees commented on pkg/build/controller/config_controller_test.go:
```diff
@@ -56,29 +90,83 @@ func TestHandleBuildConfig(t *testing.T) {
+		}
+		if tc.failFirstDelete {
+			if buildClient.deleteCount != 1 {
+				t.Errorf("%s: delete was called even though there are no builds", tc.name)
```
this error description looks wrong.

bparees commented on pkg/build/controller/config_controller_test.go:
```diff
@@ -56,29 +90,83 @@ func TestHandleBuildConfig(t *testing.T) {
+				t.Errorf("%s: delete was called even though there are no builds", tc.name)
+				continue
+			}
+			err := controller.HandleBuildConfig(tc.bc)
```
should confirm buildClient.delete gets called again?

bparees commented:
1) couple test suggestions
2) still needs @openshift/api-review  approval
3) @liggitt 's major elephant in the room.

smarterclayton commented:
Given that Brian has extant proposals to handle this upstream, that implies
we would have to change this pattern to match what upstream would pick.  I
also share concerns with Jordan.

Client side cascade for now seems like the safest option.  It's more work
for the UI, but it doesn't change the behavior of the resource to be
different from all other resources.


On Nov 24, 2015, at 4:57 PM, Jordan Liggitt <notifications@github.com>
wrote:

The current behavior of the DC controller makes oc replace --force super
dangerous. We shouldn't make another resource behave that way
unconditionally if we can help it.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5208#issuecomment-159417372>.

kargakis commented:
Agreed with @smarterclayton. The reason for choosing not to implement this in the client is an edge case anyway if I remember correctly @bparees @mnagy

mnagy commented:
OK, looks like we'll go back to the client solution. I'll find the previous solution I had with reaper and perhaps at least reuse tests from this PR.

mnagy commented:
Closing in favor of #6185
